src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -294,6 +306,42 @@ impl TargetInfo {
                 gctx.shell().warn("non-trivial mutual dependency between target-specific configuration and RUSTFLAGS")?;
             }
 
+            let mut supports_std: Option<bool> = None;
+
+            // The '--print=target-spec-json' is an unstable option of rustc, therefore only
+            // try to fetch this information if rustc allows nightly features. Additionally,
+            // to avoid making two rustc queries when not required, only try to fetch the
+            // target-spec when the '-Zbuild-std' option is passed.
+            if gctx.nightly_features_allowed && gctx.cli_unstable().build_std.is_some() {
+                let mut target_spec_process = rustc.workspace_process();
+                apply_env_config(gctx, &mut target_spec_process)?;
+                target_spec_process
+                    .arg("--print=target-spec-json")
+                    .arg("-Zunstable-options")
+                    .args(&rustflags)
+                    .env_remove("RUSTC_LOG");
+
+                if let CompileKind::Target(target) = kind {
+                    target_spec_process
+                        .arg("--target")
+                        .arg(target.rustc_target());
+                }
+
+                let target_spec_cached_output =
+                    rustc.cached_output(&target_spec_process, extra_fingerprint);
+
+                if target_spec_cached_output.is_ok() {
+                    let (stdout, _stderr) = target_spec_cached_output.unwrap();
+                    match serde_json::from_str(&stdout) {
+                        Result::Ok(val) => {
+                            let target_spec: TargetSpec = val;
+                            supports_std = target_spec.metadata.std;
+                        }
+                        Result::Err(_) => (),
+                    }
+                }
```

Comment:
`rustc.cached_output` caches the rustc output to `target/.rustc_info.json` file. I guess it is better not to cache it. A direct call to `ProcessBuilder::output` is probably good enough. Would be something like:

```suggestion
                #[derive(Deserialize)]
                pub struct Metadata {
                    pub std: Option<bool>,
                }

                #[derive(Deserialize)]
                pub struct TargetSpec {
                    pub metadata: Metadata,
                }

                if let Ok(output) = target_spec_process.output() {
                    if let Ok(spec) = serde_json::from_slice::<TargetSpec>(&output.stdout) {
                        supports_std = spec.metadata.std;
                    }
                }
```